drivers: uart-emul: Support for async API

rvoell-gcx opened the pull request:
Add support for the async uart API, so the uart emulation driver can now be used to test code using that API.

rvoell-gcx committed:
driver: uart: emul: Add support for async UART API

As it is up to a driver implementation if a partial filled buffer is
released, this behavior can be configured during runtime to allow testing
for this edge case.

Signed-off-by: Ruben Völl <ruben.voell@grandcentrix.net>

dcpleung commented:
Please fix the compliance check errors.

rvoell-gcx commented:
> Please fix the compliance check errors.

Yes, i'm already working on it

dcpleung commented on tests/drivers/uart/uart_emul/prj.conf:
```diff
@@ -1,3 +1,5 @@
+CONFIG_EVENTS=y
 CONFIG_SERIAL=y
 CONFIG_UART_INTERRUPT_DRIVEN=y
+CONFIG_UART_ASYNC_API=y
```
It would be better to create another test in testcase.yaml so both async and non-async APIs are tested.

rvoell-gcx replied:
It just enables the async API. As far as i understand it, the APIs can coexist. Otherwise the test cases would fail, as each test case checks one of the APIs. Or am i missing something?

dcpleung replied:
I still think it is a good idea to test without enabling the API at all.

rvoell-gcx replied:
I did split this into three test configurations

dcpleung commented on tests/drivers/uart/uart_async_api/src/test_uart.h:
```diff
@@ -27,4 +27,9 @@
 #define UART_NODE DT_CHOSEN(zephyr_console)
 #endif
 
+#if DT_NODE_HAS_COMPAT(UART_NODE, zephyr_uart_emul)
+#include <zephyr/drivers/serial/uart_emul.h>
+#define DUT_IS_UART_EMUL 1
+#endif
```
This doesn't seem right. It is possible that there is emulated UART, but the test is actually running with a real UART on hardware.

rvoell-gcx replied:
Thats why i checked if `UART_NODE`, which is the one used in the testcases, is the emulated uart.

bjarki-trackunit replied:
This should not be needed, as the emulated device driver should behave like any other hardware device, if the test suite must be changed to support the emulated uart, it is not emulating a uart device properly :)

rvoell-gcx replied:
I removed this. The special case for the abort test is still there, though

bjarki-trackunit commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -66,6 +66,10 @@ static void uart_async_test_init(void)
 	k_sem_reset(&rx_buf_coherency);
 	k_sem_reset(&rx_buf_released);
 	k_sem_reset(&rx_disabled);
+#ifdef DUT_IS_UART_EMUL
+	uart_emul_flush_rx_data(uart_dev);
+	uart_emul_flush_tx_data(uart_dev);
+#endif /* DUT_IS_UART_EMUL */
```
The emulated driver should behave like a real uart, I don't expect any changes should be needed for the uart test suite. as this is a clear indication that the emulated uart is not behaving as expected :)

On real hardware, the following calls should clear the rx and tx data respectively
```
	uart_rx_disable(uart_dev);
	uart_tx_abort(uart_dev);
```

rvoell-gcx replied:
I removed this, was no longer necessary anyways.

bjarki-trackunit commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -584,7 +588,16 @@ ZTEST_USER(uart_async_read_abort, test_read_abort)
 
 	uart_tx(uart_dev, tx_buf, 95, 100 * USEC_PER_MSEC);
 
+#ifndef DUT_IS_UART_EMUL
 	k_timer_start(&read_abort_timer, K_USEC(300), K_NO_WAIT);
+#else
+	/* The uart emulation driver needs no real time to send and receive data. Any waiting here
+	 * will lead to the driver being finished with sending and receiving. So instead of using
+	 * the timer here, just disable receiving before the driver had any chance to start with
+	 * processing its tasks.
+	 */
+	uart_rx_disable(uart_dev);
+#endif
```
This is also an indication that the emulated UART needs to be changed to behave like a real UART. Use a delayable work item to simulate receiving and transmitting taking time :)

rvoell-gcx replied:
First of all, thank you for your review, i really appreciate this!

But I do not agree here. I feel that simulating the timing of each individual byte being sent or received is quite a bit of overkill and would add quite a bit of complexity to this. One could calculate when the events would arrive, but this feels even more complex. I think not having the correct timing here is acceptable, even when this means that the testcase needs a small change as a compromise.

When i think about, what i want to achieve as a developer when using this fixture, i usually want to test if my code did use the API correctly and does work with the events correctly. So the events need to come in in the correct order, have the correct payloads etc. Timing is not that big of an issue, i think. E.g. if one needs the code under test to receive a message at an exact point in time, one could simply provide the data with `uart_emul_put_rx_data()` at the correct moment in time.

Also the already existing code does not do that, the implementations of the polling and the interrupt based APIs do not respect accurate timing either.

bjarki-trackunit replied:
That sounds more like a mock than an emulated UART, something better implemented using the FFF framework I think. Emulated device drivers are typically used to test the test suites themselves, and to provide a "fake" device that can be used to develop applications, like the emulated RTC and GNSS device drivers.

If the emulated UART is meant to be used as a mock, which it seems like it is, it should not be tested as part of the test suite for real UARTs, as it is not one (nor emulating one)

I would simply remove the emulated UART from the uart async api test suite, WDYT?

rvoell-gcx replied:
Thats the other extreme. I would like to have it behave like a real UART  at the API level as much as possible and the already existing test cases come in handy to test this, as those test zephyrs expectation on how the API should behave. I also got to jump through some extra hoops to make it to comply with those test cases as much as possible. So i think its a bit more than a simple mock. I just don't wanna over engineer it.

rvoell-gcx replied:
Not sure how to resolve this. I would be quite unhappy with removing it from this tests and i also do not want add simulation of the timings here, as this would add quite a bit of complexity and needs a bit of work to do for unclear benefit.

bjarki-trackunit replied:
The benefit is avoiding violating the UART API and test suite by coupling them to a specific device driver which is not behaving like a real UART, which is what the API and test suite are designed to abstract and validate.

Yes it is more complexity to make it behave like a real UART, but that in itself is not a sufficient argument against adding said complexity if the goal is to make it emulate a real UART. If the goal is not to emulate a real UART, don't add the additional complexity, and don't treat it like a real UART by for example adding it to the UART test suite :)

I'm not saying which is the better option, but those are the options AFAIK

dcpleung replied:
I do think it is better not to modify the tests to suit a particular driver.

rvoell-gcx replied:
I dropped the commit enabling these tests now.

I'm not very happy about this, as i saw some value in validating the API as far as possible. But i do not have the time to write a complete simulation of a uart peripheral. I wrote this driver as part of a project to have a nice fixture for testing code using the uart async API and thought it might be useful for others, too. I think testing that the API does help to increase confidence in those tests.

bjarki-andreasen reviewed (changes requested):
Looks really promising, a few requests from my side

rvoell-gcx commented on drivers/serial/uart_emul.c:
```diff
@@ -48,6 +54,33 @@ struct uart_emul_data {
+	bool rx_stopping;
+	bool rx_release_on_timeout;
+
+	struct uart_emul_work tx_work;
+	struct uart_emul_work rx_work;
+	struct uart_emul_work rx_disable_work;
+	struct uart_emul_delayable_work rx_timeout_work;
```
I adapted here to the "style" that was already present in the file. Is it okay if i fix this for the already existing code in this file, too? I would like to add a new commit for this before my current commits.

bjarki-trackunit replied:
I think it would be great if you updated the existing code in a commit as part of this PR :)

rvoell-gcx replied:
I refactored this, hopefully its now prettier :)

rvoell-gcx committed:
drivers: uart: emul: Remove uart_emul_work struct

This structure is bad style and not needed. Instead get the surrounding
`struct uart_emul_data` with `CONTAINER_OF()` directly.

Signed-off-by: Ruben Völl <ruben.voell@grandcentrix.net>

rvoell-gcx commented on drivers/serial/uart_emul.c:
```diff
@@ -331,6 +364,485 @@ static int uart_emul_irq_update(const struct device *dev)
+							  .offset = rx_buf_offset,
+							  .len = rx_buf_data_len,
+						  },
+				  });
```
I fixed this

rvoell-gcx commented on tests/drivers/uart/uart_async_api/testcase.yaml:
```diff
@@ -110,3 +110,8 @@ tests:
+    harness: ztest
+    platform_allow:
+      - qemu_x86
+      - native_posix
```
Using native_sim instead of native_posix now. I also moved the devicetree overlays into one special for the uart emulation node, which is then used for both platforms.

dcpleung commented on tests/drivers/uart/uart_emul/src/main.c:
```diff
@@ -9,25 +9,30 @@
 #include <zephyr/ztest.h>
 
-#define EMUL_UART_NODE	       DT_NODELABEL(euart0)
+#define EMUL_UART_NODE         DT_NODELABEL(euart0)
```
Please remove the whitespace change. This is not related.

rvoell-gcx replied:
oops, i fixed that

dcpleung commented on tests/drivers/uart/uart_emul/src/main.c:
```diff
@@ -9,25 +9,30 @@
  * tx ready IRQ event.
  */
-#define SAMPLE_DATA_SIZE       MIN(EMUL_UART_RX_FIFO_SIZE, EMUL_UART_TX_FIFO_SIZE) - 1
+#define SAMPLE_DATA_SIZE MIN(EMUL_UART_RX_FIFO_SIZE, EMUL_UART_TX_FIFO_SIZE) - 1
```
Ditto on removing whitespace changes.

rvoell-gcx committed:
tests: drivers: uart_emul: extend test case for async API

- Tests that the API of the uart-emul also works with the async API
- Tests that `uart_emul_set_release_buffer_on_timeout()` configures the
  behavior as expected
- The tests are now splited into three different configurations: only
  polling API enabled, additionally interrupt driven API enabled and
  additionally async API enabled
- Enabled `native_sim` board

Signed-off-by: Ruben Völl <ruben.voell@grandcentrix.net>

ajarmouni-st commented on tests/drivers/uart/uart_emul/testcase.yaml:
```diff
@@ -5,4 +5,22 @@ common:
+    extra_configs:
+      - CONFIG_EVENTS=y
+      - CONFIG_UART_ASYNC_API=y
+    extra_args: DTC_OVERLAY_FILE="uart_emul.overlay"
```
You could move this under `common`.

ajarmouni-st replied:
the same for
```
    platform_allow:
      - qemu_x86
      - native_sim
```

rvoell-gcx replied:
Thank you for your review. I did that.

bjarki-andreasen reviewed (approved):
Nice work! You could extend the testcase to build for native_sim_64 and qemu_x86_64 to build and run it for 64-bit archs as well :)
